File: WPF/MVVM/SingleTaskViewModel.vb
Namespace Common.MVVM

    Public MustInherit Class SingleTaskViewModel
        Inherits ViewModel

        Public Sub New(ByVal KsApplication As KsApplication)
            MyBase.New(KsApplication)
        End Sub

        Public Sub New()
            MyBase.New()
        End Sub

        Protected Friend Overrides Sub OnNavigatedTo(E As NavigationEventArgs)
            If E.NavigationType = NavigationType.NewNavigation Then
                Me.IsWorkDone = False
            End If
        End Sub

        Protected Friend Overrides Sub OnNavigatedFrom(E As NavigationEventArgs)
            If E.NavigationType = NavigationType.NewNavigation Then
                Me.IsWorkDone = True
            End If
        End Sub

        Public Function WhenWorkDone() As Task
            If Me.WhenWorkDoneTaskSource Is Nothing Then
                Me.WhenWorkDoneTaskSource = New TaskCompletionSource(Of Void)()
                If Me.IsWorkDone Then
                    Me.WhenWorkDoneTaskSource.SetResult(Nothing)
                End If
            End If

            Return Me.WhenWorkDoneTaskSource.Task
        End Function

#Region "IsWorkDone Property"
        Private _IsWorkDone As Boolean

        Public Property IsWorkDone As Boolean
            Get
                Return Me._IsWorkDone
            End Get
            Protected Set(ByVal Value As Boolean)
                If Me.SetProperty(Me._IsWorkDone, Value) Then
                    If Value Then
                        Me.WhenWorkDoneTaskSource?.SetResult(Nothing)
                    Else
                        Me.WhenWorkDoneTaskSource = Nothing
                    End If
                End If
            End Set
        End Property
#End Region

        Private WhenWorkDoneTaskSource As TaskCompletionSource(Of Void)

    End Class

End Namespace

File: WPF/MVVM/SingleTaskViewModel.cs
using System.Threading.Tasks;
using System;

namespace Ks
{
    namespace Ks.Common.MVVM
    {
        public abstract class SingleTaskViewModel : ViewModel
        {
            public SingleTaskViewModel(KsApplication KsApplication) : base(KsApplication)
            {
            }

            public SingleTaskViewModel() : base()
            {
            }

            protected internal override void OnNavigatedTo(NavigationEventArgs E)
            {
                if ((int)E.NavigationType == (int)NavigationType.NewNavigation)
                    this.IsWorkDone = false;
            }

            protected internal override void OnNavigatedFrom(NavigationEventArgs E)
            {
                if ((int)E.NavigationType == (int)NavigationType.NewNavigation)
                    this.IsWorkDone = true;
            }

            public Task WhenWorkDone()
            {
                if (this.WhenWorkDoneTaskSource == null)
                {
                    this.WhenWorkDoneTaskSource = new TaskCompletionSource<Void>();
                    if (this.IsWorkDone)
                        this.WhenWorkDoneTaskSource.SetResult(null);
                }

                return this.WhenWorkDoneTaskSource.Task;
            }

            private bool _IsWorkDone;

            public bool IsWorkDone
            {
                get
                {
                    return this._IsWorkDone;
                }
                protected set
                {
                    if (this.SetProperty(ref this._IsWorkDone, value))
                    {
                        if (value)
                            this.WhenWorkDoneTaskSource?.SetResult(null);
                        else
                            this.WhenWorkDoneTaskSource = null;
                    }
                }
            }

            private TaskCompletionSource<Void> WhenWorkDoneTaskSource;
        }
    }
}
